File: lintcode/python/0692_sliding_window_unique_elements_sum.py
#hash and sliding window

class Solution:
    """
    @param nums: the given array
    @param k: the window size
    @return: the sum of the count of unique elements in each window
    """
    def slidingWindowUniqueElementsSum(self, nums, k):
        # write your code here
        if not nums:
            return 0
        
        n = len(nums)
        sum_count = 0
        res = 0
        counter = collections.defaultdict(int)

        if k > len(nums):
            k = len(nums)

        for i in range(k):
            counter[nums[i]] += 1
            if counter[nums[i]] == 1:
                sum_count += 1

            if counter[nums[i]] == 2:
                sum_count -= 1

        res += sum_count

        left, right = 0, k
        while right < n:
   
            counter[nums[right]] += 1
            if counter[nums[right]] == 1:
                sum_count += 1

            if counter[nums[right]] == 2:
                sum_count -= 1
            
          
            counter[nums[left]] -= 1
            if counter[nums[left]] == 1:
                sum_count += 1

            if counter[nums[left]] == 0:
                sum_count -= 1

            right += 1
            left += 1

            res += sum_count

        return res

File: lintcode/golang/0692_sliding_window_unique_elements_sum.go
//hash and sliding window


/**
 * @param nums: the given array
 * @param k: the window size
 * @return: the sum of the count of unique elements in each window
 */
func slidingWindowUniqueElementsSum (nums []int, k int) int {
    // write your code here
    n := len(nums)

    if n < k {
        k = n
    }
    d := make(map[int]int)
    ret := 0
    sumCnt := 0

    for i := 0; i < k; i++ {
        d[nums[i]]++
        if d[nums[i]] == 1 {
            sumCnt++
        }
        if d[nums[i]] == 2 {
            sumCnt--
        }
    }

    ret += sumCnt

    l, r := 0, k
    for r < n {
        d[nums[r]]++
        if d[nums[r]] == 1 {
            sumCnt++
        }
        if d[nums[r]] == 2 {
            sumCnt--
        }

        d[nums[l]]--
        if d[nums[l]] == 1 {
            sumCnt++
        }
        if d[nums[l]] == 0 {
            sumCnt--
        }
        ret += sumCnt
        r++
        l++
    }

    return ret
}


// this solution time limit exceeded

/**
 * @param nums: the given array
 * @param k: the window size
 * @return: the sum of the count of unique elements in each window
 */
func slidingWindowUniqueElementsSum (nums []int, k int) int {
    // write your code here
    n := len(nums)

    if n <= k {
        k = n
    }
    ret := 0

    for i := 0; i < n-k+1; i++ {
        subNums := nums[i:i+k]
        d := make(map[int]int)
        for _, num := range subNums {
            d[num]++
        }
        for _, v := range d {
            if v == 1 {
                ret++
            }
        }
    }

    return ret
}

